### ISSUE TITLE
Create base classes for distributions to define their support

### ISSUE THREAD

**Issue by NeilGirdhar** [scipy/scipy#6038]
Given the discussion in #4598, a solution is needed to repair problems with applying location and scale parameters to distributions on the circle.  However, this is a more general problem with distributions whose support is no the real line.

The fact of scipy distributions being location scale families is a design flaw in many cases: location is meaningless for Dirichlet, Beta, Bernoulli, etc.; scale is redundant for distributions that have a concentration (e.g., Von-Mises) and meaningless for distributions that have bounded support (e.g., Dirichlet), and when you add the Von-Mises Fisher distribution on the sphere, scale can only confuse users (all samples are on the unit sphere, so what should scale mean?)

Therefore, I propose having a base class for each of the kinds of support:  the positive real line (e.g., gamma), the real line (e.g., normal), the positive integers (e.g., Poisson), the circle (e.g., Von-Mises), the unit sphere in n-dimensions (e.g., Von-Mises Fisher), the simplex in n dimensions (e.g. beta, Dirichlet), bounded real-valued support (e.g., uniform), bounded integer-valued support (e.g., uniform, categorical, multinomial, binomial).

Distributions with bounded support should probably have neither location nor scale.  Only distributions on the real line should probably have location.

**Comment by josef-pkt** [comment]
> The fact of scipy distributions being location scale families is a design flaw in many cases

No it's not!

Because of an asymmetry in usage and implementation. We can always set location to zero (which has also been supported in `fit` for several years) if location is implemented. But we cannot add location if it is not implemented.

How do you decide which cases to rule out. E.g. 
Beta or other distribution with support on an interval [a, b]  can easily be shifted to any distribution in an arbitrary interval [min, max].
Pareto or variants of it is a popular case for shifting location so we model only the upper tail of a distribution (e.g. peak over threshold). 

scipy.stats has one discrete distribution defined on all integers (including negative), the finite support discrete distribution can have arbitrary points (?). Truncated Poisson or Truncated NegativeBinomial would have support starting at some integer > 0.

Truncated continuous distribution like normal could be defined on any interval.

Then there are a few distribution where the support depends on the parameter and is moving around as function of the shape parameters.

The big advantage is that the user can reuse the current implementation in various ways, with a bit of work. If the restrictions are hard coded, then no alternative use is possible.

circular or spherical is different, because we have to wrap or map back into the right space, whether that is in degrees, hours of the day or day of the year or radians or whatever.

**Comment by josef-pkt** [comment]
Note, I'm all in favor of #4538 like paramnormal as wrappers, but I wouldn't like it if my favorite unusual use breaks because of hardcoding more decisions.
(I'm arguing theoretically, I don't think I or statsmodels has a case where some of the usage would break with some of the restrictions. The only common usecases that I know for shifting lower bound are in extreme value applications eg. the pareto case.)

**Comment by NeilGirdhar** [comment]
> The fact of scipy distributions being location scale families is a design flaw in many cases

> No it's not!

I'm sorry, but it is in the case of exponential families.  Did you read my arguments?  It would be helpful if you would explain what a scale applied to a Von-Mises Fisher distribution means? 

Shifting the beta distribution is in my opinion almost never what you want to do.  If you want to work with a distribution over an arbitrary interval having a sample mean and variance, you almost always actually want a truncated normal distribution.  Nevertheless, there's nothing wrong with having an adaptor for a distribution with bounded support that shifts and scales it.  Similarly, why not have an adaptor to all distributions that truncate their support?

In other words, I'm not suggesting hard-coding a restriction.  What I'm suggesting is coding the meaning of the support so that appropriate adaptors can be made:
* a scaling adaptor (e.g., it would scale down the pdf)
* a translation adaptor
* a truncation adaptor (e.g., it would have to calculate the cdf over the truncated region and use that as the normalizer when calculating the pdf)

**In short, some of the complexity that is currently baked into distributions (location and scale) should actually be moved to adaptors.  Some distributions that have a location or scale simply shouldn't.**

**Comment by josef-pkt** [comment]
Now we get into the implementation issue

If you have a simplifying proposal, then it would be good to look at. There are more developers watching this that it might be worth it.

The current situation is that the base classes are these "wrappers". The distribution specific methods `_xxx` don't know anything about loc and scale. 
This provided a single access point for this functionality and works well generically, with headaches for a few distributions. It was enough work to get this to work properly initially. 

Splitting up the base classes multiplies the code and maintenance, especially if you still want additional wrappers for loc scale transformation for restricted distributions.
We don't want to leave coding of those helper functions to users, because it's still error prone. (e.g. we initially a mistake in loc scale handling in the `expect` function. Treating fixed parameter in `fit` is not trivial at all.)

My impression is that writing wrappers around the current generic classes as is done in paramnormal is less work and creates fewer sources of error because it essentially just needs to be a collection of transformation (reparameterization) functions.

The main advantage I see is if we can reduce some of the generic overhead in the computation. I sometimes use the distribution specific `_xxx` methods for speed, but I'm usually not interested enough to reimplement loc and scale for a specific usecase. It's easier just to use the provided methods.

**Comment by NeilGirdhar** [comment]
> The current situation is that the base classes are these "wrappers". The distribution specific methods _xxx don't know anything about loc and scale. 

Right, and this is exactly the problem.   For a support [-pi, pi] representing a circle, a scale is dangerously wrong, and the location needs to be interpreted correctly.  This is why this functionality should be put into a subclass (_separation of concerns_).  One idea is to make subclasses for each of the support types and implement location and scale appropriately.  For a circular distribution on [-pi, pi], no scale would be provided.  For a circular distribution expressed as a unit vector in 2d, scale is fine.

**Comment by NeilGirdhar** [comment]
@mdhaber We can close this, right?  Looks like you have a `support` method.  (Although, I'm not sure what you're going to do for circular distributions though.)

Also, very cool that you ended up doing my adapter idea for scale and transformation (and much more).

**Comment by mdhaber** [comment]
If it's satisfied from your perspective, certainly. It looks like we don't have classes for some of these things, but we will keep you in the loop for circular and whenever we start thinking about flavors of multivariate.

**Comment by NeilGirdhar** [comment]
Okay, I'll just leave it open after circular distributions are done, at which point it should be clearer what you've chosen for their support.  But I'm happy there's a method now.
